classic/selection/FeatureModel.js

```diff
@@ -0,0 +1,238 @@
+/* Copyright (c) 2015-2019 The Open Source Geospatial Foundation
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/**
+  * A row selection model which enables automatic selection of features
+  * in the map when rows are selected in the grid and vice-versa.
+  *
+  * @class GeoExt.selection.FeatureModel
+  */
+Ext.define('GeoExt.selection.FeatureModel', {
+    extend: 'Ext.selection.RowModel',
+    alias: 'selection.featuremodel',
+
+    /**
+     * The currently selected features.
+     * @property {ol.Collection<ol.Feature>}
+     */
+    selectedFeatures: null,
+
+    /**
+     * The connected vector layer.
+     * @cfg {ol.layer.Vector}
+     * @property {ol.layer.Vector}
+     */
+    layer: null,
+
+    /**
+     * The OpenLayers map we work on
+     * @cfg {ol.Map}
+     */
+    map: null,
+
+    /**
+     * Set to true to create a click handler on the map selecting a clicked
+     * object in the #layer.
+     * @cfg {Boolean}
+     */
+    mapSelection: false,
+
+    /**
+     * The default style for the selected features.
+     * @cfg {ol.style.Style}
+     */
+    selectStyle: new ol.style.Style({
+        image: new ol.style.Circle({
+            radius: 6,
+            fill: new ol.style.Fill({
+                color: 'rgba(255,255,255,0.8)'
+            }),
+            stroke: new ol.style.Stroke({
+                color: 'darkblue',
+                width: 2
+            })
+        }),
+        fill: new ol.style.Fill({
+            color: 'rgba(255,255,255,0.8)'
+        }),
+        stroke: new ol.style.Stroke({
+            color: 'darkblue',
+            width: 2
+        })
+    }),
+
+    /**
+     * The attribute key to mark an OL feature as selected.
+     * @cfg {String}
+     * @property
+     * @readonly
+     */
+    selectedFeatureAttr: 'gx_selected',
+
+    /**
+     * Prepare several connected objects once the selection model is ready.
+     *
+     * @private
+     */
+    bindComponent: function() {
+        var me = this;
+
+        me.callParent(arguments);
+
+        me.selectedFeatures = new ol.Collection();
+
+        // detect a layer from the store if not passed in
+        if (!me.layer || !me.layer instanceof ol.layer.Vector) {
+            var store = me.getStore();
+            if (store && store.getLayer() &&
```

Comment:
Is getLayer() here to check if this function exists? Could a store.getLayer be used instead? This was throwing errors with a bound store (as it is defaults to a "ext-empty-store" when first created before the store in the ViewModel is bound.

Reply:
We should test both: `store.getLayers` and `store.getLayer()`. Done in 1eada2c8df8ec3231e22097630f8edd7722403dc